packages/vite/src/node/ssr/runtimeClient.ts

```diff
@@ -0,0 +1,558 @@
+          callstack[callstack.length - 2]
+        }`,
+      )
+
+    const modulePath = cleanUrl(moduleId)
+    // disambiguate the `<UNIT>:/` on windows: see nodejs/node#31710
+    const href = pathToFileURL(modulePath).href
+    const meta = { url: href }
+    const exports = Object.create(null)
+    Object.defineProperty(exports, Symbol.toStringTag, {
+      value: 'Module',
+      enumerable: false,
+      configurable: false,
+    })
+    Object.defineProperty(exports, '__esModule', { value: true })
+    // this proxy is triggered only on exports.{name} and module.exports access
+    // inside the module itself. imported module is always "exports"
+    const cjsExports = new Proxy(exports, {
+      get: (target, p, receiver) => {
+        if (Reflect.has(target, p)) return Reflect.get(target, p, receiver)
+        return Reflect.get(Object.prototype, p, receiver)
+      },
+      getPrototypeOf: () => Object.prototype,
+      set: (_, p, value) => {
+        // treat "module.exports =" the same as "exports.default =" to not have nested "default.default",
+        // so "exports.default" becomes the actual module
+        if (
+          p === 'default' &&
+          this.shouldInterop(modulePath, { default: value })
+        ) {
+          exportAll(cjsExports, value)
+          exports.default = value
+          return true
+        }
+
+        if (!Reflect.has(exports, 'default')) exports.default = {}
+
+        // returns undefined, when accessing named exports, if default is not an object
+        // but is still present inside hasOwnKeys, this is Node behaviour for CJS
+        if (isPrimitive(exports.default)) {
+          defineExport(exports, p, () => undefined)
+          return true
+        }
+
+        exports.default[p] = value
+        if (p !== 'default') defineExport(exports, p, () => value)
+
+        return true
+      },
+    })
+
+    Object.assign(mod, { code: transformed, exports })
+
+    const __filename = fileURLToPath(href)
+    const moduleProxy = {
+      set exports(value) {
+        exportAll(cjsExports, value)
+        exports.default = value
+      },
+      get exports() {
+        return cjsExports
+      },
+    }
+
+    // Vite hot context
+    let hotContext: ViteHotContext | undefined
+    if (this.options.createHotContext) {
+      Object.defineProperty(meta, 'hot', {
+        enumerable: true,
+        get: () => {
+          hotContext ||= this.options.createHotContext?.(
+            (id) => this.executeId(id),
+            `/@fs/${fsPath}`,
+          )
+          return hotContext
+        },
+        set: (value) => {
+          hotContext = value
+        },
+      })
+    }
+
+    // Be careful when changing this
+    // changing context will change amount of code added on line :114 (vm.runInThisContext)
+    // this messes up sourcemaps for coverage
+    // adjust `offset` variable in packages/coverage-c8/src/provider.ts#86 if you do change this
+    const context = this.prepareContext({
```

Comment:
The length of this execution wrapper should be exported. It could be a constant number like `VITE_NODE_WRAPPER_LENGTH` or a function that can generate the wrapper code, `function wrapCode(code: string)`.

Then we'll use it directly in Vitest: https://github.com/vitest-dev/vitest/blob/158c4bb056417828833088508b43e6917d65d23f/packages/coverage-v8/src/provider.ts#L40-L41. It's then used by [v8-to-istanbul as `wrapperLength`](https://github.com/istanbuljs/v8-to-istanbul/blob/master/lib/v8-to-istanbul.js).